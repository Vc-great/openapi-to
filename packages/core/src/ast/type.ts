import type { JSDocTagStructure } from "ts-morph";
import type {
  AmbientableNodeStructure,
  ExportableNodeStructure,
  JSDocableNodeStructure,
  Structure,
} from "ts-morph";
import type {
  BindingNamedNodeStructure,
  ExclamationTokenableNodeStructure,
  KindedStructure,
  OptionalKind,
  StructureKind,
  TypedNodeStructure,
  VariableDeclarationKind,
  WriterFunction,
} from "ts-morph";

export interface VariableStatementStructureOfObject
  extends Structure,
    VariableStatementSpecificStructure,
    JSDocableNodeStructure,
    AmbientableNodeStructure,
    ExportableNodeStructure {}

interface VariableStatementSpecificStructure
  extends KindedStructure<StructureKind.VariableStatement> {
  declarationKind?: VariableDeclarationKind;
  declarations: OptionalKind<VariableDeclarationStructure>[];
}

export interface VariableDeclarationStructure
  extends Structure,
    VariableDeclarationSpecificStructure,
    BindingNamedNodeStructure,
    InitializerExpressionableNodeStructure,
    TypedNodeStructure,
    ExclamationTokenableNodeStructure {}

interface VariableDeclarationSpecificStructure
  extends KindedStructure<StructureKind.VariableDeclaration> {}

export interface InitializerExpressionableNodeStructure {
  initializer?: string | WriterFunction | ObjectStructure[];
}

export interface ObjectStructure {
  key: string;
  value: string;
  docs: JSDocStructure[];
}

export interface JSDocStructure {
  /**
   * The description of the JS doc.
   * @remarks To force this to be multi-line, add a newline to the front of the string.
   */
  description?: string;
  /** JS doc tags (ex. `&#64;param value - Some description.`). */
  tags?: OptionalKind<JSDocTagStructure>[];
}

export type JSDoc = JSDocStructure[];
