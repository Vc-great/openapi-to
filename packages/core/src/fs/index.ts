export { clean } from './clean.ts'
export { pathExistsSync } from './pathExistsSync.ts'
export { read, readSync } from './read.ts'
export { write } from './write.ts'
