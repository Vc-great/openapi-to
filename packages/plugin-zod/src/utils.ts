import _ from 'lodash'

import { ZOD_NAME_SPACE_SUFFIX, ZOD_SUFFIX, ZOD_TYPE_MODEL_SUFFIX } from './constants.ts'

export function formatRefName(ref: string): string {
  return _.lowerFirst(ref) + _.upperFirst(ZOD_SUFFIX)
}

export function formatZodNameSpaceName(zodName: string): string {
  return `${_.lowerFirst(zodName) + _.upperFirst(ZOD_NAME_SPACE_SUFFIX)}`
}

export function formatZodName(zodName: string): string {
  return _.lowerFirst(zodName) + _.upperFirst(ZOD_SUFFIX)
}

export function formatZodTypeName(zodName: string): string {
  return _.upperFirst(zodName) + _.upperFirst(ZOD_TYPE_MODEL_SUFFIX)
}

export function formatFileName(fileName: string): string {
  return `${_.lowerFirst(fileName)}.${ZOD_SUFFIX}`
}

export function formatZodNameSpaceFileName(fileName: string): string {
  return `${_.lowerFirst(fileName)}.${ZOD_SUFFIX}s`
}
