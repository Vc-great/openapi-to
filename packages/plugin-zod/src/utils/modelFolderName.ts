export const modelFolderName = "zod-schema";
