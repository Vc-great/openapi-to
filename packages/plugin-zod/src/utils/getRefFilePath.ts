import { getRefAlias } from '@openapi-to/core/utils'
import { kebabCase } from 'lodash-es'

export function getRefFilePath(ref: string, dir: string) {
  const str = ref.split('/')[2]
  const fileName = kebabCase(getRefAlias(ref))

  switch (str) {
    case 'parameters':
      return `${dir}/parameters/${fileName}.schema.ts`
    case 'requestBodies':
      return `${dir}/requestBodies/${fileName}.schema.ts`
    case 'responses':
      return `${dir}/responses/${fileName}.schema.ts`
    case 'schemas':
      return `${dir}/models/${fileName}.schema.ts`
    default:
      return ''
  }
}
