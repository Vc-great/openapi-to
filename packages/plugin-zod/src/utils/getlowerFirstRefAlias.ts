import { getRefAlias } from '@openapi-to/core/utils'
import { lowerFirst, nth, split } from 'lodash-es'

export function getlowerFirstRefAlias($ref: string) {
  const alias: Record<string, string> = {
    parameters: 'parameter',
    requestBodies: 'requestBodies',
    responses: 'responses',
  }
  const prefix = nth(split($ref, '/'), 2) || ''

  const prefixAlias = prefix in alias ? alias[prefix] : ''
  return `${lowerFirst(getRefAlias($ref))}Schema`
}
