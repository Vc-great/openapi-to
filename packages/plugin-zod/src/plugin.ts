import { buildSchemaImports } from '@/builds/buildSchemaImports.ts'

import { buildComponentParameters } from '@/builds/components/buildComponentParameters.ts'
import { buildComponentsRequestBody } from '@/builds/components/buildComponentsRequestBody.ts'
import { buildComponentsResponse } from '@/builds/components/buildComponentsResponse.ts'

import { buildSchemas } from '@/builds/components/buildSchemas.ts'

import { collectRefsFromComponentParameters, collectRefsFromComponentRequestBody, collectRefsFromComponentResponse } from '@/collect/collectRefsFromDocument.ts'
import { collectRefsFromOperation } from '@/collect/collectRefsFromOperation.ts'
import { collectRefsFromSchema } from '@/collect/collectRefsFromSchemas.ts'
import { importZodTemplate } from '@/templates/importZodTemplate.ts'
import { getOperationZodSchemaName } from '@/templates/operationTypeNameTemplate.ts'
import { getlowerFirstRefAlias } from '@/utils/getlowerFirstRefAlias.ts'

import { getRefFilePath } from '@/utils/getRefFilePath.ts'
import { createPlugin, pluginEnum } from '@openapi-to/core'
import { formatterModuleSpecifier } from '@openapi-to/core/utils'
import { getRelativePath } from '@openapi-to/core/utils'
import { forEach, kebabCase, upperFirst } from 'lodash-es'

import path from 'node:path'
import { Project } from 'ts-morph'
import { buildOperationTypes } from './builds/buildOperationTypes.ts'
import type { PluginConfig } from './types.ts'

const schemaFolderName = 'zod'

export const definePlugin = createPlugin((pluginConfig?: PluginConfig) => {
  const project = new Project()

  const operationFileNameOfTag: Set<string> = new Set()
  let componentOutputDir = ''
  return {
    name: pluginEnum.Zod,
    hooks: {
      buildStart: async (ctx) => {
        componentOutputDir = path.join(ctx.openapiToSingleConfig.output.dir, schemaFolderName)
      },
      tagStart: async (tagData, ctx) => {
        operationFileNameOfTag.clear()
      },
      operation: async (operation, ctx) => {
        const fileName = `${kebabCase(operation.accessor.operationName)}.schema.ts`
        const filePath = path.join(ctx.openapiToSingleConfig.output.dir, kebabCase(operation.tagName), fileName)
        operationFileNameOfTag.add(fileName)

        const operationStatements = buildOperationTypes(operation)

        //
        operation.accessor.setOperationZodSchemaName({
          ...getOperationZodSchemaName(operation),
          filePath,
        })

        const operationSourceFile = project.createSourceFile(filePath, '', { overwrite: true })

        const imports = collectRefsFromOperation(operation).flatMap((ref) => {
          return buildSchemaImports(
            [getlowerFirstRefAlias(ref)],
            formatterModuleSpecifier(getRelativePath(filePath, getRefFilePath(ref, componentOutputDir)), pluginConfig?.importWithExtension),
          )
        })

        operationSourceFile.addStatements([...imports, importZodTemplate, ...operationStatements])
        ctx.setSourceFiles([pluginEnum.Zod, operation.tagName, operation.accessor.operationName], operationSourceFile)
      },
      tagEnd: async (tagData, ctx) => {},
      componentsSchemas: async (schemas, ctx) => {
        for (const schemaName in schemas) {
          const schema = schemas[schemaName]!

          const formatterSchemaName = ctx.openapiHelper.formatterName(schemaName)

          const fileName = `${kebabCase(formatterSchemaName)}.schema.ts`

          const filePath = path.join(componentOutputDir, 'models', fileName)

          const schemaSourceFile = project.createSourceFile(filePath, '', { overwrite: true })
          const statements = buildSchemas(formatterSchemaName, schema)
          const refs = collectRefsFromSchema(schema)

          const imports = refs.flatMap((ref) =>
            buildSchemaImports(
              [getlowerFirstRefAlias(ref)],
              formatterModuleSpecifier(getRelativePath(filePath, getRefFilePath(ref, componentOutputDir)), pluginConfig?.importWithExtension),
            ),
          )

          statements && schemaSourceFile.addStatements([...imports, importZodTemplate, ...(statements ? [statements] : [])])

          ctx.setSourceFiles([pluginEnum.Zod, 'componentsSchemas', schemaName], schemaSourceFile)
        }
      },
      componentsParameters(parameters, ctx) {
        const refs = collectRefsFromComponentParameters(parameters)

        forEach(parameters, (parameter, parameterName) => {
          const formatterParameterName = ctx.openapiHelper.formatterName(parameterName)

          const fileName = `${kebabCase(formatterParameterName)}.schema.ts`

          const filePath = path.join(componentOutputDir, 'parameters', fileName)
          const parameterSourceFile = project.createSourceFile(filePath, '', {
            overwrite: true,
          })
          const statements = buildComponentParameters(parameter, formatterParameterName)
          if (!statements) {
            return
          }
          const imports = refs.flatMap((ref) =>
            buildSchemaImports(
              [getlowerFirstRefAlias(ref)],
              formatterModuleSpecifier(getRelativePath(filePath, getRefFilePath(ref, componentOutputDir)), pluginConfig?.importWithExtension),
            ),
          )

          parameterSourceFile.addStatements([...imports, importZodTemplate, statements])

          ctx.setSourceFiles([pluginEnum.Zod, 'componentsParameters', parameterName], parameterSourceFile)
        })
      },
      componentsRequestBodies(requestBodies, ctx) {
        // components.requestBodies
        for (const [requestBodyName, requestObject] of Object.entries(requestBodies)) {
          const formatterName = ctx.openapiHelper.formatterName(requestBodyName)

          const refs = collectRefsFromComponentRequestBody(requestObject)

          const fileName = `${kebabCase(formatterName)}.schema.ts`

          const filePath = path.join(componentOutputDir, 'requestBodies', fileName)
          const requestBodySourceFile = project.createSourceFile(filePath, '', {
            overwrite: true,
          })
          const statements = buildComponentsRequestBody(formatterName, requestObject)
          if (!statements) {
            return
          }
          const imports = refs.flatMap((ref) =>
            buildSchemaImports(
              [getlowerFirstRefAlias(ref)],
              formatterModuleSpecifier(getRelativePath(filePath, getRefFilePath(ref, componentOutputDir)), pluginConfig?.importWithExtension),
            ),
          )

          requestBodySourceFile.addStatements([...imports, importZodTemplate, statements])
          ctx.setSourceFiles([pluginEnum.Zod, 'componentsRequestBodies', requestBodyName], requestBodySourceFile)
        }
      },
      componentsResponses(responses, ctx) {
        // components.responses
        forEach(responses, (response, responseName) => {
          const formatterResponse = ctx.openapiHelper.formatterName(responseName)

          const responseTypeName = `Response${upperFirst(formatterResponse)}`
          //todo responses
          const statements = buildComponentsResponse(response, responseTypeName)

          const refs = collectRefsFromComponentResponse(response)

          const fileName = `${kebabCase(formatterResponse)}.schema.ts`

          const filePath = path.join(componentOutputDir, 'responses', fileName)

          const responseSourceFile = project.createSourceFile(filePath, '', {
            overwrite: true,
          })

          if (!statements) {
            return
          }
          const imports = refs.flatMap((ref) =>
            buildSchemaImports(
              [getlowerFirstRefAlias(ref)],
              formatterModuleSpecifier(getRelativePath(filePath, getRefFilePath(ref, componentOutputDir)), pluginConfig?.importWithExtension),
            ),
          )

          responseSourceFile.addStatements([...imports, importZodTemplate, statements])
          ctx.setSourceFiles([pluginEnum.Zod, 'componentsResponses', responseName], responseSourceFile)
        })
      },
      buildEnd: async (ctx) => {},
    },
  }
})
