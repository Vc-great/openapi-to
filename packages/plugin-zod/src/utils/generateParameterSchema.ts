import { jsDocTemplateFromParameter, writeJSDoc } from '@/templates/jsDocTemplateFromSchema.ts'
import { resolveBaseSchema } from '@/templates/schemaTemplate.ts'

import { getlowerFirstRefAlias } from '@/utils/getlowerFirstRefAlias.ts'

import type { ParameterObjectWithRef } from '@openapi-to/core'
import type { OptionalKind, PropertySignatureStructure } from 'ts-morph'
import { CodeBlockWriter } from 'ts-morph'

type OptionalKindOfPropertySignatureStructure = OptionalKind<PropertySignatureStructure>

/**
 * 将一组属性描述生成成带 JSDoc 注释的 TypeScript 对象字面量
 */
export function generateParameterSchema(parameters: ParameterObjectWithRef[], operationName: string): string {
  const writer = new CodeBlockWriter({ indentNumberOfSpaces: 4 })

  writer.block(() => {
    parameters.forEach((prop, idx) => {
      const name = prop.name
      const comma = idx < parameters.length - 1 ? ',' : ''

      if ('$ref' in prop && prop.$ref) {
        const refSchemaName = getlowerFirstRefAlias(prop.$ref)
        writer.writeLine(`${name}: ${refSchemaName}${comma}`)
      } else {
        writeJSDoc(writer, jsDocTemplateFromParameter(prop))
        const optional = prop.required ? '' : '.optional()'
        const schemaString = prop.schema ? resolveBaseSchema(prop.schema, name, operationName) : '.string()'

        writer.writeLine(`${name}: z${schemaString}${optional}${comma}`)
      }
    })
  })
  return `z.object(${writer.toString()})`
}
