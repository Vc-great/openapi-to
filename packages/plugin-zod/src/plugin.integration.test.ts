import fs from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'
import { PluginManager } from '@openapi-to/core'
import { checkFolderHasFiles } from '@openapi-to/core/utils'

import type { MediaTypeObject, SchemaObject } from 'oas/types'
import { afterEach, beforeEach, describe, expect, it } from 'vitest'
import mockOpenAPI from '../mock/petstore.json'
import { definePlugin } from './plugin'
// 获取当前文件的目录
const __dirname = path.dirname(fileURLToPath(import.meta.url))
const TEST_OUTPUT_DIR = path.resolve(__dirname, '../test-output')

// 模拟简单的 OpenAPI 操作
const mockOperation = {
  accessor: {
    operationName: 'getUser',
    operation: {
      getResponseStatusCodes: () => ['200', '404'],
      getResponseAsJSONSchema: () => [
        {
          description: 'User response',
          label: 'user',
          schema: { type: 'object', properties: { id: { type: 'string' } } },
          type: 'object',
        },
      ],
      getRequestBody: (mediaType?: string): MediaTypeObject | false | [string, MediaTypeObject, ...string[]] => {
        return {
          schema: {
            type: 'object',
            properties: {
              id: { type: 'string' },
              name: { type: 'string' },
            },
          },
        }
      },
    },
    getParametersSchema(): SchemaObject {
      return {
        type: 'object',
        properties: {
          id: { type: 'string', description: 'Unique identifier' },
          name: { type: 'string', description: 'Name of the entity' },
        },
        required: ['id'],
      }
    },
  },
}

describe('Ts Type Plugin Integration', () => {
  beforeEach(() => {
    // 确保测试输出目录存在
    if (!fs.existsSync(TEST_OUTPUT_DIR)) {
      fs.mkdirSync(TEST_OUTPUT_DIR, { recursive: true })
    }

    // 确保 types 目录存在
    const typesDir = path.join(TEST_OUTPUT_DIR, 'types')
    if (!fs.existsSync(typesDir)) {
      fs.mkdirSync(typesDir, { recursive: true })
    }
  })

  afterEach(() => {
    // 清理测试文件
    const typesDir = path.join(TEST_OUTPUT_DIR, 'types')
    if (fs.existsSync(typesDir)) {
      const files = fs.readdirSync(typesDir)
      files.forEach((file) => {
        // fs.unlinkSync(path.join(typesDir, file))
      })
    }
  })

  it('should generate typescript file with correct namespace', async () => {
    const pluginManager = new PluginManager(
      {
        root: '',
        plugins: [definePlugin()],
        input: {
          path: '',
        },
        output: {
          dir: TEST_OUTPUT_DIR,
        },
      },
      // @ts-ignore
      mockOpenAPI,
    )

    pluginManager.run()

    // 检查文件是否生成
    const hasFiles = checkFolderHasFiles(TEST_OUTPUT_DIR)
    expect(hasFiles).toBe(true)

    // 读取文件内容并进行简单验证
    // const fileContent = fs.readFileSync(filePath, 'utf-8')
    // expect(fileContent).toContain('namespace Users')

    expect('').toMatchSnapshot()
  })
})
