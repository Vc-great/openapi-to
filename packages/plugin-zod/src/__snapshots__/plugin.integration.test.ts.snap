// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Ts Type Plugin Integration > should generate typescript file with correct namespace 1`] = `""`;
